pp/series_data/decoder/interval_decode_iterator.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "series_data/encoder/sample.h"
+#include "traits.h"
+
+namespace series_data::decoder {
+
+template <class DecodeIterator, class DecodeIteratorSentinel>
+class IntervalDecodeIterator : public DecodeIteratorTypeTrait {
+ public:
+  using Timestamp = PromPP::Primitives::Timestamp;
+
+  IntervalDecodeIterator(DecodeIterator&& iterator, DecodeIteratorSentinel&& end, Timestamp interval, Timestamp lookback)
+      : iterator_(std::move(iterator)), iterator_end_(std::move(end)), interval_(std::max(interval, kMinInterval)), lookback_(lookback) {
+    if (iterator_ != iterator_end_) {
+      timestamp_ = align_timestamp(iterator_->timestamp, interval_);
+      advance_to_next_sample();
+    }
+  }
+
+  PROMPP_ALWAYS_INLINE const encoder::Sample& operator*() const noexcept { return sample_; }
+  PROMPP_ALWAYS_INLINE const encoder::Sample* operator->() const noexcept { return &sample_; }
+
+  PROMPP_ALWAYS_INLINE bool operator==(const DecodeIteratorSentinel&) const noexcept { return iterator_ == iterator_end_ && sample_.timestamp == kNoSample; }
+
+  PROMPP_ALWAYS_INLINE IntervalDecodeIterator& operator++() noexcept {
+    advance_to_next_sample();
+    return *this;
+  }
+
+  PROMPP_ALWAYS_INLINE IntervalDecodeIterator operator++(int) noexcept {
+    const auto result = *this;
+    ++*this;
+    return result;
+  }
+
+ private:
+  static constexpr auto kNoSample = std::numeric_limits<Timestamp>::min();
+  static constexpr Timestamp kMinInterval = 1;
+
+  DecodeIterator iterator_;
+  [[no_unique_address]] DecodeIteratorSentinel iterator_end_;
+  Timestamp interval_;
+  Timestamp lookback_;
+  Timestamp timestamp_{};
+  encoder::Sample sample_{.timestamp = kNoSample};
+
+  PROMPP_ALWAYS_INLINE static Timestamp align_timestamp(Timestamp timestamp, Timestamp align_to) noexcept {
```

Comment:
IMHO "align" isn't clear meaning word. In this case more mathematically correct is "round_up_to_step"